source/droid/gateway/gateway.d

```diff
@@ -131,7 +146,44 @@ final class Gateway
         }
 
         logger_.infof("Lost connection, close code %d (reason %s)", ws_.closeCode, ws_.closeReason);
-        exitEventLoop(true);
+
+        // User-initiated shutdown.
+        if (shuttingDown) {
+            exitEventLoop(true);
+            return;
+        }
+
+        reconnectionAttempts++;
+
+        if (reconnectionAttempts > MAX_RECONNECTS) {
+            logger_.errorf("Failed to reconnect too many times, quitting...");
+            exitEventLoop(true);
+            return;
+        }
+
+        // They keep trying, it's not working out.
+        if (reconnectionAttempts > 1) {
+            sessionId_ = null;
+            lastSeqNum_ = 0;
+            logger_.errorf("Reconnecting in 5 seconds..");
+
+            sleep(5.seconds);
```

Comment:
You should wait at a growing backoff with some maximum, and remove `MAX_RECONNECTS`

https://github.com/meew0/discordcr/blob/master/src/discordcr/client.cr#L133-L144